pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/BrokerRegistryImpl.java

```diff
@@ -0,0 +1,218 @@
+import org.apache.pulsar.broker.PulsarService;
+import org.apache.pulsar.broker.ServiceConfiguration;
+import org.apache.pulsar.broker.loadbalance.extensions.data.BrokerLookupData;
+import org.apache.pulsar.common.util.FutureUtil;
+import org.apache.pulsar.metadata.api.MetadataStoreException;
+import org.apache.pulsar.metadata.api.Notification;
+import org.apache.pulsar.metadata.api.NotificationType;
+import org.apache.pulsar.metadata.api.coordination.LockManager;
+import org.apache.pulsar.metadata.api.coordination.ResourceLock;
+
+/**
+ * The broker registry impl, base on the LockManager.
+ */
+@Slf4j
+public class BrokerRegistryImpl implements BrokerRegistry {
+
+    private static final String LOOKUP_DATA_PATH = "/loadbalance/brokers";
+
+    private final PulsarService pulsar;
+
+    private final ServiceConfiguration conf;
+
+    private final BrokerLookupData brokerLookupData;
+
+    private final LockManager<BrokerLookupData> brokerLookupDataLockManager;
+
+    private final String brokerZNodePath;
+
+    private final String lookupServiceAddress;
+
+    @VisibleForTesting
+    protected final Map<String, BrokerLookupData> brokerLookupDataMap;
+
+    private final ScheduledExecutorService scheduler;
+
+    private final List<BiConsumer<String, NotificationType>> listeners;
+
+    private final AtomicBoolean registered;
+
+    private volatile ResourceLock<BrokerLookupData> brokerLookupDataLock;
+
+    public BrokerRegistryImpl(PulsarService pulsar) {
+        this.pulsar = pulsar;
+        this.conf = pulsar.getConfiguration();
+        this.brokerLookupDataLockManager = pulsar.getCoordinationService().getLockManager(BrokerLookupData.class);
+        this.scheduler = pulsar.getLoadManagerExecutor();
+        this.brokerLookupDataMap = new ConcurrentHashMap<>();
+        this.listeners = new ArrayList<>();
+
+        this.registered = new AtomicBoolean(false);
+        this.lookupServiceAddress = pulsar.getAdvertisedAddress() + ":"
+                + conf.getWebServicePort().orElseGet(() -> conf.getWebServicePortTls().get());
+        this.brokerZNodePath = LOOKUP_DATA_PATH + "/" + lookupServiceAddress;
+        this.brokerLookupData = new BrokerLookupData(
+                pulsar.getSafeWebServiceAddress(),
+                pulsar.getWebServiceAddressTls(),
+                pulsar.getBrokerServiceUrl(),
+                pulsar.getBrokerServiceUrlTls(),
+                pulsar.getAdvertisedListeners(),
+                pulsar.getProtocolDataToAdvertise(),
+                pulsar.getConfiguration().isEnablePersistentTopics(),
+                pulsar.getConfiguration().isEnableNonPersistentTopics(),
+                pulsar.getBrokerVersion());
+    }
+
+    @Override
+    public void start() {
+        pulsar.getLocalMetadataStore().registerListener(this::handleMetadataStoreNotification);
+    }
+
+    @Override
+    public void register() {
+        if (registered.compareAndSet(false, true)) {
```

Comment:
what if `brokerLookupDataLockManager .acquireLock(..)` fails (but we pre-set `registered` to true in this case)?